File: 백준/Platinum/3015. 오아시스 재결합/오아시스 재결합.cc
#include <iostream>
#include <stack>
#define fastio cin.tie(0), cout.tie(0), ios::sync_with_stdio(0);
using namespace std;


stack <pair <int, int>> st;
int a[501010];

int main(){
    fastio;
    int n; cin >> n;
    long long result = 0;
    for(int i = 1;i <= n;i++) cin >> a[i];

    for(int i = 1;i <= n;i++){
        while(!st.empty() && st.top().first < a[i]){
            result += st.top().second; st.pop();
        }

        if(st.empty() || st.top().first != a[i]) st.push({a[i], 1});
        else{
            int num = st.top().second + 1;
            st.pop(); st.push({a[i], num});
            result += num - 1;
        }
        
        if(st.size() >= 2) result++;
    }

    cout << result;
    return 0;
}

File: 백준/Platinum/3015. 오아시스 재결합/오아시스 재결합.py
import sys
input = sys.stdin.readline
n = int(input().rstrip())
a = [int(input().rstrip()) for _ in range(n)]

st = []
result = 0

def comp():
    if len(st) <= 1:
        return
    
    top = st.pop()
    if st[-1][0] != top[0]:
        st.append(top)
        return

    num = st[-1][1] + top[1]
    st.pop()
    st.append((top[0], num))

for i in a:
    while len(st) and st[-1][0] < i:
        result += st[-1][1]
        st.pop()

    if len(st) == 0 or st[-1][0] != i:
        st.append((i, 1))
    else:
        num = st[-1][1] + 1
        st.pop()
        st.append((i, num))
        result += num - 1

    comp()

    if(len(st) >= 2):
        result += 1

print(result) 